File: packages/app-core/platforms/android/app/src/main/java/ai/elizaos/app/AgentPlugin.java
package ai.elizaos.app;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;

@CapacitorPlugin(name = "Agent")
public class AgentPlugin extends Plugin {
    private static final String LOCAL_AGENT_BASE_URL = "http://127.0.0.1:31337";
    private static final int DEFAULT_TIMEOUT_MS = 10_000;
    private static final int MIN_TIMEOUT_MS = 1_000;
    private static final int MAX_TIMEOUT_MS = 120_000;
    private static final int MAX_REQUEST_BODY_BYTES = 10 * 1024 * 1024;
    private static final int MAX_RESPONSE_BODY_BYTES = 10 * 1024 * 1024;

    @PluginMethod
    public void start(PluginCall call) {
        try {
            ElizaAgentService.start(getContext());
            call.resolve(status("starting", null));
        } catch (Exception error) {
            call.reject(error.getMessage() != null ? error.getMessage() : "Failed to start local agent");
        }
    }

    @PluginMethod
    public void stop(PluginCall call) {
        try {
            ElizaAgentService.stop(getContext());
            JSObject result = new JSObject();
            result.put("ok", true);
            call.resolve(result);
        } catch (Exception error) {
            call.reject(error.getMessage() != null ? error.getMessage() : "Failed to stop local agent");
        }
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        final String token = readLocalAgentToken();
        if (token == null) {
            call.resolve(status("not_started", null));
            return;
        }

        new Thread(() -> {
            try {
                JSObject result = forwardLocalRequest("/api/status", "GET", new JSObject(), null, 1500, token);
                String body = result.getString("body", "{}");
                JSONObject json = new JSONObject(body);
                String state = json.optString("state", "running");
                String error = json.optString("error", "").trim();
                call.resolve(status(state, error.isEmpty() ? null : error));
            } catch (Exception error) {
                call.resolve(status("error", error.getMessage() != null ? error.getMessage() : "Local agent status unavailable"));
            }
        }, "eliza-agent-status").start();
    }

    @PluginMethod
    public void getLocalAgentToken(PluginCall call) {
        String token = readLocalAgentToken();
        JSObject result = new JSObject();
        result.put("available", token != null);
        result.put("token", token != null ? token : JSONObject.NULL);
        call.resolve(result);
    }

    @PluginMethod
    public void request(PluginCall call) {
        String path = trimToNull(call.getString("path"));
        if (path == null || !path.startsWith("/") || path.startsWith("//") || hasUrlScheme(path)) {
            call.reject("Agent.request requires a local path that starts with / and is not an absolute URL");
            return;
        }

        String requestedMethod = trimToNull(call.getString("method"));
        final String method = requestedMethod == null
            ? "GET"
            : requestedMethod.toUpperCase(Locale.US);
        if (!method.matches("^[A-Z]{1,16}$")) {
            call.reject("Unsupported HTTP method");
            return;
        }

        int timeoutMs = clampTimeout(call.getInt("timeoutMs", DEFAULT_TIMEOUT_MS));
        String body = call.getString("body");
        JSObject headers = call.getObject("headers", new JSObject());
        String token = readLocalAgentToken();

        new Thread(() -> {
            try {
                JSObject result = forwardLocalRequest(path, method, headers, body, timeoutMs, token);
                call.resolve(result);
            } catch (Exception error) {
                call.reject(error.getMessage() != null ? error.getMessage() : "Local agent request failed");
            }
        }, "eliza-agent-request").start();
    }

    private static JSObject status(String state, String error) {
        JSObject result = new JSObject();
        result.put("state", state);
        result.put("agentName", JSONObject.NULL);
        result.put("port", "not_started".equals(state) ? JSONObject.NULL : 31337);
        result.put("startedAt", JSONObject.NULL);
        result.put("error", error != null ? error : JSONObject.NULL);
        return result;
    }

    private static String readLocalAgentToken() {
        String token = ElizaAgentService.localAgentToken();
        if (token == null) return null;
        String trimmed = token.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private static JSObject forwardLocalRequest(
        String path,
        String method,
        JSObject headers,
        String body,
        int timeoutMs,
        String token
    ) throws Exception {
        byte[] requestBody = body != null ? body.getBytes(java.nio.charset.StandardCharsets.UTF_8) : null;
        if (requestBody != null && requestBody.length > MAX_REQUEST_BODY_BYTES) {
            throw new IllegalArgumentException("Request body is too large");
        }

        HttpURLConnection connection = (HttpURLConnection) new URL(LOCAL_AGENT_BASE_URL + path).openConnection();
        connection.setRequestMethod(method);
        connection.setConnectTimeout(timeoutMs);
        connection.setReadTimeout(timeoutMs);
        connection.setInstanceFollowRedirects(false);
        connection.setUseCaches(false);

        applyHeaders(connection, headers);
        if (token != null && trimToNull(connection.getRequestProperty("Authorization")) == null) {
            connection.setRequestProperty("Authorization", "Bearer " + token);
        }

        if (requestBody != null && !"GET".equals(method) && !"HEAD".equals(method)) {
            connection.setDoOutput(true);
            connection.getOutputStream().write(requestBody);
        }

        int status = connection.getResponseCode();
        String responseBody = readResponseBody(connection, status);
        JSObject responseHeaders = new JSObject();
        for (Map.Entry<String, List<String>> entry : connection.getHeaderFields().entrySet()) {
            String key = entry.getKey();
            List<String> values = entry.getValue();
            if (key == null || values == null || values.isEmpty()) continue;
            responseHeaders.put(key.toLowerCase(Locale.US), String.join(", ", values));
        }

        JSObject result = new JSObject();
        result.put("status", status);
        result.put("statusText", connection.getResponseMessage() != null ? connection.getResponseMessage() : "");
        result.put("headers", responseHeaders);
        result.put("body", responseBody);
        return result;
    }

    private static void applyHeaders(HttpURLConnection connection, JSObject headers) {
        Iterator<String> keys = headers.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            if (key == null || isBlockedHeader(key)) continue;
            Object value = headers.opt(key);
            if (!(value instanceof String)) continue;
            String trimmed = trimToNull((String) value);
            if (trimmed != null) {
                connection.setRequestProperty(key, trimmed);
            }
        }
    }

    private static String readResponseBody(HttpURLConnection connection, int status) throws Exception {
        InputStream stream = status >= 400 ? connection.getErrorStream() : connection.getInputStream();
        if (stream == null) return "";
        try (InputStream input = stream; ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[8192];
            int total = 0;
            while (true) {
                int count = input.read(buffer);
                if (count == -1) break;
                total += count;
                if (total > MAX_RESPONSE_BODY_BYTES) {
                    throw new IllegalStateException("Response body is too large");
                }
                output.write(buffer, 0, count);
            }
            return output.toString(java.nio.charset.StandardCharsets.UTF_8.name());
        }
    }

    private static boolean isBlockedHeader(String key) {
        return "host".equalsIgnoreCase(key)
            || "connection".equalsIgnoreCase(key)
            || "content-length".equalsIgnoreCase(key);
    }

    private static int clampTimeout(Integer value) {
        int timeout = value != null ? value : DEFAULT_TIMEOUT_MS;
        return Math.min(MAX_TIMEOUT_MS, Math.max(MIN_TIMEOUT_MS, timeout));
    }

    private static boolean hasUrlScheme(String path) {
        return path.matches("^[a-zA-Z][a-zA-Z0-9+.-]*://.*");
    }

    private static String trimToNull(String value) {
        if (value == null) return null;
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}

File: packages/native-plugins/agent/android/src/main/java/ai/eliza/plugins/agent/AgentPlugin.kt
package ai.eliza.plugins.agent

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale
import org.json.JSONObject

private const val LOCAL_AGENT_BASE_URL = "http://127.0.0.1:31337"
private const val MAX_REQUEST_BODY_BYTES = 10 * 1024 * 1024
private const val MAX_RESPONSE_BODY_BYTES = 10 * 1024 * 1024
private const val DEFAULT_REQUEST_TIMEOUT_MS = 10_000
private const val MAX_REQUEST_TIMEOUT_MS = 600_000
private const val STARTUP_POLL_INTERVAL_MS = 250L
private const val STARTUP_TOKEN_WAIT_MS = 30_000L
private const val STARTUP_REQUEST_RETRY_MS = 60_000L

/**
 * Eliza Agent Plugin — Android bridge.
 *
 * The app module owns ElizaAgentService, so this library uses reflection to
 * avoid a Gradle dependency cycle while still exposing the per-boot loopback
 * bearer token to the WebView.
 */
@CapacitorPlugin(name = "Agent")
class AgentPlugin : Plugin() {
    @PluginMethod
    fun start(call: PluginCall) {
        try {
            invokeAgentService("start")
            call.resolve(agentStatus("starting", null))
        } catch (error: Exception) {
            call.reject(error.message ?: "Failed to start local agent")
        }
    }

    @PluginMethod
    fun stop(call: PluginCall) {
        try {
            invokeAgentService("stop")
            call.resolve(JSObject().apply {
                put("ok", true)
            })
        } catch (error: Exception) {
            call.reject(error.message ?: "Failed to stop local agent")
        }
    }

    @PluginMethod
    fun getStatus(call: PluginCall) {
        val token = readLocalAgentToken()
        if (token == null) {
            call.resolve(agentStatus("not_started", null))
            return
        }

        Thread {
            try {
                val result = forwardLocalRequest("/api/status", "GET", JSObject(), null, 1_500, token)
                val httpStatus = result.getInt("status")
                if (httpStatus < 200 || httpStatus >= 300) {
                    call.resolve(agentStatus("starting", "Local agent status HTTP $httpStatus"))
                    return@Thread
                }
                val json = JSONObject(result.getString("body") ?: "{}")
                call.resolve(agentStatus(
                    json.optString("state", "running"),
                    json.optString("error").takeIf { it.isNotBlank() },
                ))
            } catch (error: Exception) {
                val message = error.message ?: "Local agent status unavailable"
                val state = if (isTransientLoopbackStartupError(error)) "starting" else "error"
                call.resolve(agentStatus(state, message))
            }
        }.start()
    }

    @PluginMethod
    fun getLocalAgentToken(call: PluginCall) {
        val token = readLocalAgentToken()
        call.resolve(JSObject().apply {
            put("available", token != null)
            put("token", token ?: JSONObject.NULL)
        })
    }

    @PluginMethod
    fun request(call: PluginCall) {
        val path = call.getString("path")?.trim()
        if (path == null || !path.startsWith("/") || path.startsWith("//")) {
            call.reject("Agent.request requires a local path that starts with /")
            return
        }

        val method = (call.getString("method") ?: "GET").trim().uppercase(Locale.US)
        if (!method.matches(Regex("^[A-Z]{1,16}$"))) {
            call.reject("Unsupported HTTP method")
            return
        }

        val timeoutMs = (call.getInt("timeoutMs") ?: DEFAULT_REQUEST_TIMEOUT_MS)
            .coerceIn(1_000, MAX_REQUEST_TIMEOUT_MS)
        val body = call.getString("body")
        val headers = call.getObject("headers") ?: JSObject()

        Thread {
            try {
                val token = ensureLocalAgentToken(timeoutMs)
                val result = forwardLocalRequestWithStartupRetry(path, method, headers, body, timeoutMs, token)
                call.resolve(result)
            } catch (error: Exception) {
                call.reject(error.message ?: "Local agent request failed")
            }
        }.start()
    }

    private fun agentStatus(state: String, error: String?): JSObject {
        return JSObject().apply {
            put("state", state)
            put("agentName", JSONObject.NULL)
            put("port", if (state == "not_started") JSONObject.NULL else 31337)
            put("startedAt", JSONObject.NULL)
            put("error", error ?: JSONObject.NULL)
        }
    }

    private fun invokeAgentService(methodName: String) {
        val serviceClass = resolveAgentServiceClass()
        val method = serviceClass.getMethod(methodName, android.content.Context::class.java)
        method.invoke(null, context)
    }

    private fun resolveAgentServiceClass(): Class<*> {
        val packageNames = listOfNotNull(
            context.packageName,
            context.applicationContext?.packageName,
            "ai.elizaos.app",
        ).map { it.trim() }
            .filter { it.isNotEmpty() }
            .distinct()

        var lastError: ClassNotFoundException? = null
        for (packageName in packageNames) {
            try {
                return Class.forName("$packageName.ElizaAgentService")
            } catch (error: ClassNotFoundException) {
                lastError = error
            }
        }
        throw lastError ?: ClassNotFoundException("ElizaAgentService")
    }

    private fun isTransientLoopbackStartupError(error: Exception): Boolean {
        if (error is IOException) return true
        val message = error.message?.lowercase(Locale.US) ?: return false
        return message.contains("connection refused") ||
            message.contains("failed to connect") ||
            message.contains("timed out") ||
            message.contains("unexpected end of stream")
    }

    private fun readLocalAgentToken(): String? {
        return try {
            val serviceClass = resolveAgentServiceClass()
            val method = serviceClass.getMethod("localAgentToken")
            (method.invoke(null) as? String)?.trim()?.takeIf { it.isNotEmpty() }
        } catch (_: Exception) {
            null
        }
    }

    private fun ensureLocalAgentToken(timeoutMs: Int): String? {
        readLocalAgentToken()?.let { return it }
        invokeAgentService("start")

        val waitMs = minOf(timeoutMs.toLong(), STARTUP_TOKEN_WAIT_MS)
        val deadline = System.currentTimeMillis() + waitMs
        while (System.currentTimeMillis() < deadline) {
            readLocalAgentToken()?.let { return it }
            try {
                Thread.sleep(STARTUP_POLL_INTERVAL_MS)
            } catch (error: InterruptedException) {
                Thread.currentThread().interrupt()
                throw IOException("Interrupted while waiting for local agent token", error)
            }
        }
        return readLocalAgentToken()
    }

    private fun forwardLocalRequestWithStartupRetry(
        path: String,
        method: String,
        headers: JSObject,
        body: String?,
        timeoutMs: Int,
        initialToken: String?,
    ): JSObject {
        var token = initialToken
        val retryMs = minOf(timeoutMs.toLong(), STARTUP_REQUEST_RETRY_MS)
        val deadline = System.currentTimeMillis() + retryMs

        while (true) {
            try {
                return forwardLocalRequest(path, method, headers, body, timeoutMs, token)
            } catch (error: Exception) {
                if (!isTransientLoopbackStartupError(error)) {
                    throw error
                }
                if (System.currentTimeMillis() >= deadline) {
                    throw error
                }
                token = token ?: readLocalAgentToken()
                try {
                    Thread.sleep(STARTUP_POLL_INTERVAL_MS)
                } catch (interrupted: InterruptedException) {
                    Thread.currentThread().interrupt()
                    throw IOException("Interrupted while waiting for local agent", interrupted)
                }
            }
        }
    }

    private fun forwardLocalRequest(
        path: String,
        method: String,
        headers: JSObject,
        body: String?,
        timeoutMs: Int,
        token: String?,
    ): JSObject {
        val requestBody = body?.toByteArray(Charsets.UTF_8)
        if (requestBody != null && requestBody.size > MAX_REQUEST_BODY_BYTES) {
            throw IllegalArgumentException("Request body is too large")
        }

        val connection = (URL("$LOCAL_AGENT_BASE_URL$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = timeoutMs
            readTimeout = timeoutMs
            instanceFollowRedirects = false
            useCaches = false
        }

        for (key in headers.keys()) {
            if (key.equals("host", ignoreCase = true) ||
                key.equals("connection", ignoreCase = true) ||
                key.equals("content-length", ignoreCase = true) ||
                key.equals("authorization", ignoreCase = true)
            ) {
                continue
            }
            val value = headers.opt(key) as? String
            if (!value.isNullOrBlank()) {
                connection.setRequestProperty(key, value)
            }
        }

        if (token != null) {
            connection.setRequestProperty("Authorization", "Bearer $token")
        }

        if (requestBody != null && method != "GET" && method != "HEAD") {
            connection.doOutput = true
            connection.outputStream.use { output ->
                output.write(requestBody)
            }
        }

        val status = connection.responseCode
        val stream = if (status >= 400) connection.errorStream else connection.inputStream
        val responseBody = stream?.use { input ->
            val output = ByteArrayOutputStream()
            val buffer = ByteArray(8192)
            var total = 0
            while (true) {
                val count = input.read(buffer)
                if (count == -1) break
                total += count
                if (total > MAX_RESPONSE_BODY_BYTES) {
                    throw IllegalStateException("Response body is too large")
                }
                output.write(buffer, 0, count)
            }
            output.toString(Charsets.UTF_8.name())
        } ?: ""

        val responseHeaders = JSObject()
        for ((key, values) in connection.headerFields) {
            if (key == null || values == null || values.isEmpty()) continue
            responseHeaders.put(key.lowercase(Locale.US), values.joinToString(", "))
        }

        return JSObject().apply {
            put("status", status)
            put("statusText", connection.responseMessage ?: "")
            put("headers", responseHeaders)
            put("body", responseBody)
        }
    }
}
